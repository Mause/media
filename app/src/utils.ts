import { String } from 'typescript-string-operations';
import { useState, useEffect } from 'react';
import type { RawAxiosRequestHeaders } from 'axios';
import Axios from 'axios';
import * as RRD from 'react-router-dom';
// import axiosRetry from '@vtex/axios-concurrent-retry';
import type {
  Auth0ContextInterface,
  AuthenticationError,
} from '@auth0/auth0-react';
import { useAuth0 } from '@auth0/auth0-react';
import moment from 'moment';
import * as _ from 'lodash-es';

import { FetchEventTarget } from './fetch_stream';
import type { TV } from './SeasonSelectComponent';
import type { EpisodeResponse } from './streaming';

// axiosRetry(Axios, { retries: 3 });

export type GetResponse<T> = T extends {
  get: { responses: { '200': { content: { 'application/json': unknown } } } };
}
  ? T['get']['responses']['200']['content']['application/json']
  : never;

export function subscribe<T>(
  path: string,
  callback: (a: T) => void,
  error: (e: Error) => void,
  authorization: string,
  end: (() => void) | null = () => null,
): () => void {
  const es = FetchEventTarget(path, {
    headers: new Headers({
      Authorization: `Bearer ${authorization}`,
    }),
  });
  const onerror = (event: Event) => {
    error(event as unknown as Error);
  };
  es.addEventListener('abort', onerror);
  const internal_callback = (event: Event) => {
    callback((event as MessageEvent).data as T);
  };
  es.addEventListener('message', internal_callback);

  return () => {
    es.removeEventListener('close', end);
    es.removeEventListener('abort', onerror);
    es.removeEventListener('message', internal_callback);
  };
}

export function getPrefix() {
  const prefix = import.meta.env.REACT_APP_API_PREFIX;

  if (!prefix) {
    return '';
  } else if (prefix.includes('localhost')) {
    return '';
  } else if (prefix) {
    return `https://${prefix}`;
  }
}

export async function load<T>(
  path: string,
  params?: object,
  headers?: RawAxiosRequestHeaders,
): Promise<T> {
  const url = `${getPrefix()}/api/${path}`;
  console.log('Loading', { url, params });
  const t = await Axios.get<T>(url, {
    params,
    withCredentials: true,
    headers,
  });
  return t?.data;
}

interface Res<T> {
  data?: T;
  done: boolean;
  error?: Error;
}

function isAuthenticationError(e: unknown): e is AuthenticationError {
  return typeof e === 'object' && e != null && 'error' in e;
}

export async function getToken(auth0: Auth0ContextInterface): Promise<string> {
  try {
    return await auth0.getAccessTokenSilently();
  } catch (e) {
    if (isAuthenticationError(e) && e.error === 'missing_refresh_token') {
      await auth0.loginWithRedirect({
        authorizationParams: {
          redirect_uri: window.location.toString(),
        },
      });
      return '';
    } else {
      throw e;
    }
  }
}

export function usePost<T>(
  url: string,
  body: object,
): { done: boolean; data?: T; error?: Error } {
  const [result, setResult] = useState<Res<T>>({ done: false });
  const auth = useAuth0();

  useEffect(() => {
    const abortController = new AbortController();
    getToken(auth)
      .then((token) =>
        Axios.post<T>(`/api/${url}`, body, {
          signal: abortController.signal,
          headers: {
            Authorization: `Bearer ${token}`,
          },
        }),
      )
      .then(
        (res) => {
          setResult({ done: true, data: res.data });
        },
        (error: unknown) => {
          setResult({ done: true, error: error as Error });
        },
      );

    return () => {
      abortController.abort();
    };
  }, [url, body, auth]);

  return result;
}

export function useLocation<T>() {
  const location = RRD.useLocation();
  return { ...location, state: location.state as T };
}

export function getMarker(episode: {
  season?: number;
  episode?: number | null;
}) {
  if (episode.episode) {
    return String.format('S{0:00}E{1:00}', episode.season, episode.episode);
  } else {
    return String.format('S{0:00}', episode.season);
  }
}

export function getMessage(air_date: string) {
  const today = moment().startOf('day');
  const tomorrow = today.clone().add(1, 'day');
  const yesterday = today.clone().subtract(1, 'day');
  const dt = moment(air_date);
  const dts = dt.format('DD/MM/YYYY');

  let message;
  if (today.isSame(dt)) {
    message = 'airs today';
  } else if (dt.isSame(yesterday)) {
    message = 'aired yesterday';
  } else if (dt.isSame(tomorrow)) {
    message = 'airs tomorrow';
  } else if (dt.isAfter(today)) {
    message = `airs on ${dts}`;
  } else {
    message = `aired on ${dts}`;
  }
  return message;
}

export function shouldCollapse(
  i: string,
  data: TV | undefined,
  episodes: EpisodeResponse[],
): boolean {
  let collapse = false;
  if (data) {
    const i_i = +i;
    const seasonMeta = data.seasons.find((s) => s.season_number === i_i);
    if (seasonMeta) {
      const hasNext = true; // !!data.seasons[i_i + 1];

      const episodeNumbers = _.range(1, seasonMeta.episode_count + 1);
      const hasNumbers = _.map(episodes, 'episode');
      const hasAllEpisodes =
        _.difference(episodeNumbers, hasNumbers).length === 0;
      collapse = hasNext && hasAllEpisodes;
    }
  }
  return collapse;
}
