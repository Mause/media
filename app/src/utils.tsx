import React, { ReactElement } from 'react';
import Axios from 'axios';
import { useState, useEffect } from 'react';
import MaterialLink from '@material-ui/core/Link';
import { Link } from 'react-router-dom';
import { LocationDescriptor } from 'history';
import axiosRetry from '@vtex/axios-concurrent-retry';
import { TypographyTypeMap } from '@material-ui/core';
import moxios from 'moxios';
import { unstable_batchedUpdates } from 'react-dom';

// axiosRetry(Axios, { retries: 3 });

export function MLink<S>(props: {
  children: React.ReactNode;
  to: LocationDescriptor<S>;
  color?: TypographyTypeMap['props']['color'];
}): ReactElement {
  return <MaterialLink component={Link} {...props} />;
}

export function subscribe<T>(
  path: string,
  callback: (a: T) => void,
  end: (() => void) | null = null,
): void {
  const es = new EventSource(path, {
    withCredentials: true,
  });
  es.addEventListener('message', ({ data }) => {
    if (!data) {
      if (end) {
        end();
      }
      return es.close();
    }
    callback(JSON.parse(data));
  });
}

export async function load<T>(path: string, params?: string): Promise<T> {
  const t = await Axios.get<T>(`/api/${path}`, {
    params,
    withCredentials: true,
  });
  return t && t.data;
}

export function usePost<T>(
  url: string,
  body: object,
): { done: boolean; data?: T; error?: Error } {
  const [done, setDone] = useState<boolean>(false);
  const [error, setError] = useState<Error>();
  const [data, setData] = useState<T>();

  useEffect(
    () => {
      Axios.post<T>('/api/' + url, body, {
        withCredentials: true,
      }).then(
        res => {
          unstable_batchedUpdates(() => {
            setDone(true);
            setData(res.data);
          });
        },
        error => {
          unstable_batchedUpdates(() => {
            setDone(true);
            setError(error);
          });
        },
      );
    },
    [url, body],
  );

  return { done, error, data };
}

export function ExtMLink(props: { href: string; children: string }) {
  return (
    <MaterialLink href={props.href} target="_blank" rel="noopener noreferrer">
      {props.children}
    </MaterialLink>
  );
}

export function expectLastRequestBody() {
  return expect(JSON.parse(moxios.requests.mostRecent().config.data));
}
