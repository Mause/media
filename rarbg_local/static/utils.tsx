import Axios from 'axios';
import { useState, useEffect } from 'react';

export function subscribe(path: string, callback: (a: any) => void, end: (() => void) | null = null): void {
  const es = new EventSource(path, {
    withCredentials: true,
  });
  es.addEventListener('message', ({ data }) => {
    if (!data) {
      if (end) {
        end();
      }
      return es.close();
    }
    callback(JSON.parse(data));
  });
}

export function load<T>(path: string, params?: any): Promise<T> {
  return Axios.get<T>(`/api/${path}`, { params, withCredentials: true }).then(t => t.data);
}

export function useLoad<T>(path: string, params: any = null) {
  const [data, setData] = useState<T>()
  useEffect(() => {
    load<T>(path, params).then(setData);
  }, [path, params]);
  return data;
}
