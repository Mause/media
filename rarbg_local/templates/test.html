{% extends "base.html" %}

{% block header %}
Options for "{{title}}"
{% endblock %}

{% block content %}
<script type="text/javascript">
  var require = {
    bundles: JSON.parse(`{{BUNDLES|safe}}`),
    deps: ['streaming'],
    callback: function () {
      console.log('Loaded');
    }
  };
</script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.6/require.js"></script>

<div id="app"></div>
{% endblock %}
