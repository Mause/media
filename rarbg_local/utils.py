from functools import lru_cache as _lru_cache
from typing import Set

_caches: Set[_lru_cache] = set()  # type: ignore


def lru_cache(*args, **kwargs):
    def wrapper(func):
        cache = _lru_cache(*args, **kwargs)(func)
        _caches.add(cache)
        return cache

    return wrapper


def cache_clear():
    while _caches:
        cache = _caches.pop()
        cache.cache_clear()  # type: ignore
