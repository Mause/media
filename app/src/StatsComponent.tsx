import React from 'react';
import { useLoad } from './utils';
import ReactLoading from 'react-loading';

export type StatsResponse = {
  user: string;
  values: {
    movie: number;
    episode: number;
  };
};

export function StatsComponent() {
  const stats = useLoad<StatsResponse[]>('stats');

  if (!stats) return <ReactLoading type='balls' color='#000' />;

  return <div>
    {stats.map(({user, values}) => <div key={user}>
      <h3>{user}</h3>
      Movie: {values.movie || 0}<br />
      Episode: {values.episode || 0}<br />
    </div>)}
  </div>;
}
