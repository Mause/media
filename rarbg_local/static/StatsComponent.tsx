import React from 'react';
import { useLoad } from './utils';
import _ from 'lodash';

export function StatsComponent() {
  const stats = useLoad<{
    [key: string]: {
      movie: number;
      episode: number;
    };
  }>('stats');
  return <div>
    {stats ? _.map(stats, (stat, username) => <div key={username}>
      <h3>{username}</h3>
      Movie: {stat.movie || 0}<br />
      Episode: {stat.episode || 0}<br />
    </div>) : null}
  </div>;
}
