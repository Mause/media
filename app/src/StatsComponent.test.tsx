import React from 'react';

import { StatsComponent, StatsResponse } from './StatsComponent';
import { mock, usesMoxios, renderWithSWR, wait } from './test.utils';

usesMoxios();

test('render', async () => {
  const { container } = renderWithSWR(<StatsComponent />);

  await mock<StatsResponse[]>('/api/stats', [
    { user: 'Mause', values: { episode: 1, movie: 1 } },
  ]);
  await wait();

  expect(container).toMatchSnapshot();
});
