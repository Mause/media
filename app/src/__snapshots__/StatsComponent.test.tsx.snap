// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render 1`] = `
<div>
  <div>
    <div>
      <h3>
        Mause
      </h3>
      Movie: 
      1
      <br />
      Episode: 
      1
      <br />
    </div>
  </div>
</div>
`;
